lib/caching-precompiler.js

```diff
@@ -47,8 +48,25 @@ CachingPrecompiler.prototype._init = function () {
 CachingPrecompiler.prototype._transform = function (code, filePath, hash) {
 	code = code.toString();
 
-	var options = babelConfigHelper.build(this.babelConfig, this.powerAssert, filePath, code);
-	var result = this.babel.transform(code, options);
+	var fileExt = path.extname(filePath).slice(1);
+	var isCustomExt = fileExt !== 'js' && this.extensions.indexOf(fileExt) > -1;
+
+	var result;
+
+	if (isCustomExt) {
+		switch (fileExt) {
+			case 'ts':
+				var extTs = require('./ext/ts');
+				result = extTs.transpile(code, filePath);
+				break;
+
+			default:
+				throw new Error('`--extension ' + fileExt + '` is not supported by AVA.');
```

Comment:
Let's check for allowed extensions in [`lib/cli` instead](https://github.com/avajs/ava/blob/0f0400158ec4192fc32ccc603877d0b088302662/lib/cli.js#L104).